### PR TITLE
World data structure rearchitcture with possession system rebuild

### LINKED ISSUE #1361: World datastructures cleanup
### Description of the task

A high-level planning ticket to cleanup all our `world` datastructures.

#1355 and #1358 started work to separate "state" objects from versions that require time information. Additionally we have realized there are lots of places where it would be sufficient to use non-timestamped classes, as well as that lots of the "top-level" world datastructures like `Ball` are used everywhere, when they aren't necessarily needed.

We should audit where we case about
1. buffer data / history
2. Timestamped data
3. Purely "physical state" data

and try use our types more carefully. We may need to refactor certain classes.

### Acceptance criteria

- [ ] TODO after investigation

### LINKED ISSUE #1522: Move Possession, Robot History and Ball History to SensorFusion
### Description of the task
Currently past TimestampedBallStates and TimestampedRobotStates are recorded in a circular buffer in Robot and are used by evaluation functions to determine the things like possession. Since we want to incorporate sensor data, such as break beam at the sensor fusion layer, we need to move the evaluation functions down to the sensorfusion layer

### Acceptance criteria

- [ ] Move Possession, Robot History and Ball History to SensorFusion

### Blocked By
#1523

### LINKED ISSUE #1604: Create new Possession State and Add to World
### Description of the task
as titled

### Acceptance criteria

- [ ] Add classes
- [ ] Add tests

### LINKED ISSUE #1606: Integrate possession state into tactic, action, evaluation
### Description of the task
Since possession state will duplicate a lot of functionality of evaluation, we can delete many of those functions and replace them with possession state.

### Acceptance criteria

- [ ] Add possession state to updateWorldParams where needed
- [ ] Modify tests appropriately
- [ ] Replace/integrate possession evaluation functions 

### Blocked By
#1604

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
The current system of passing a `World` object with histories of `Robot`, `Ball`, etc. makes it difficult to test (mocking histories is hard) and involves significant complexity to maintain and use. With SensorFusion, we can move histories into that layer and have the same functionality or better. In this PR, we remove a lot of the heavy `World` history data structure in favour of a simplified possession system as a basis to rebuild evaluations. We can easily add back `teamPassInProgress` to the SensorFusion layer in the future if that is needed. 

* Removed ball history
* Removed robot history
* Removed world timestamp history
* Removed team timestamp history
* Removed TimestampedBallState
* Removed TimestampedRobotState

### Testing Done
blocked on #1767, but otherwise CI still passes

### Resolved Issues
- resolves #1606 
- resolves #1604 
- resolves #1522 
- resolves #1361

### Length Justification
Mainly deletions

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/world/robot.cpp [hunk 1]
```diff
@@ -60,27 +46,39 @@ RobotId Robot::id() const
...
-    return states_.front().state().angularVelocity();
+    return current_state_.angularVelocity();
 }
 
-RobotHistory Robot::getPreviousStates() const
+bool Robot::isNearDribbler(const Point &test_point) const
 {
-    return states_;
+    Vector vector_to_test_point                     = test_point - position();
+    static const double POSSESSION_THRESHOLD_METERS = ROBOT_MAX_RADIUS_METERS + 0.2;
```
Comment (kelvinkoon) [inline src/software/world/robot.cpp:71]: Is the `0.2` just from trial and error or is there a specific reason? Just curious where it came from.

File: src/software/world/robot.h [hunk 2]
```diff
@@ -132,6 +123,15 @@ class Robot
      */
     std::set<RobotCapability> &getMutableRobotCapabilities();
 
+    /**
+     * Decides if a point is near the dribbler of the robot
+     *
+     * @param test_point The point to check
+     *
+     * @return whether the test_point is near the dribbler of the robot
```
Comment (kelvinkoon) [inline src/software/world/robot.h:131]: In the `robot.c`, we define "near" by if the point is within the 90 degree cone in front of the robot. Would it better if we included this in the javadoc? Could be better than just "near the dribbler".

File: src/software/world/team_types.h [hunk 3]
```diff
@@ -2,6 +2,11 @@
 
 #include "software/util/make_enum/make_enum.h"
 
+/**
+ * The possible team side of SSL robots
```
Comment (kelvinkoon) [inline src/software/world/team_types.h]: "The possible team side**s**..." to keep it consistent with comment below.

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Yeetage of code looks great.

File: src/software/world/robot.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/world/robot.h:131]: the implementation details shouldn't be in the java doc, but I can add comments to the function body

File: src/software/world/robot.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/world/robot.cpp:71]: it's experimental, I've added a comment for that

File: src/software/world/team_types.h (hunk 3 above)
Comment (jonathanlew) [reply src/software/world/team_types.h]: fixed

File: src/software/world/robot.h (hunk 2 above)
Comment (kelvinkoon) [reply src/software/world/robot.h:131]: The function body already has a comment on it (I think) so no need.

**Review by kelvinkoon — DISMISSED** [review]

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
ezgif-3-74704ddeb951
yellow new possesion system
blue old system
this gif isn't great but figured Id toss it in because i recorded it

There seem to be _no_ visible regressions, all play transitions from refbox events seem to work, the only thing thats sus, is that we are _waaay_ to chill when we have "last touched" the ball. But that work is out of the scope of this pr, we eventually will adapt the offensive/defensive plays to try to grab the ball even if we have "team possession"

Very nice work, greatly simplifies things

File: src/software/world/robot.h [hunk 4]
```diff
@@ -42,31 +36,36 @@ class Robot
      *
      * @param id The id of the robot to create
      * @param initial_state The initial state of the robot
-     * @param history_size The number of previous robot states that should be stored. Must
-     * be > 0
+     * @param timestamp The timestamp at which the robot was observed to be in the given
+     * state
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/world/robot.h:40]: `Simulator` uses `RobotState` in a way that it doesn't care about `Timestamp`. Remember that `Timestamp` is an SSL-Vision construct

File: src/software/world/world.h [hunk 5]
```diff
@@ -40,8 +39,10 @@ class World final
      * Updates the state of the ball in the world with the new ball data
      *
      * @param new_ball_data A BallState containing new ball information
+     * @param new_timestamp New Timestamp
      */
-    void updateBallStateWithTimestamp(const TimestampedBallState& new_ball_state);
+    void updateBallStateWithTimestamp(const BallState& new_ball_state,
+                                      const Timestamp& new_timestamp);
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/world/world.h]: we don't want to expose the internal `Ball` object from `World`, but changed it to `updateBall` instead

File: src/software/world/ball.h [hunk 6]
```diff
@@ -51,19 +46,19 @@ class Ball final
...
      *
      * @param new_state the new state of the ball
+     * @param new_timestamp the new timestamp
      */
-    void updateState(const TimestampedBallState &new_state);
+    void updateState(const BallState &new_state, const Timestamp &new_timestamp);
 
     /**
-     * Returns the timestamp for when this ball's data was last updated
+     * Returns the current timestamp for when this ball
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/world/ball.h]: fixed

File: src/software/sensor_fusion/sensor_fusion.cpp [hunk 7]
```diff
@@ -178,41 +177,41 @@ void SensorFusion::updateWorld(const SSLProto::SSL_DetectionFrame &ssl_detection
...
-    {
-        ball->updateState(new_ball_state);
-    }
-    else
-    {
-        ball = Ball(new_ball_state);
+        if (!friendly_team_has_ball && enemy_team_has_ball)
+        {
+            team_with_possession = TeamSide::ENEMY;
+        }
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.cpp]: ok fixed the logic

**Review by itsarune — COMMENTED** [review]
👍 Answers the questions I had last week about Ball and TimestampedBall

**Review by akhilveeraghanta — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]

**Pull request merged**